packages/rest/src/lib/RequestManager.ts

```diff
@@ -0,0 +1,276 @@
+		const headers: RequestHeaders = {
+			'User-Agent': `${DefaultUserAgent} ${options.userAgentAppendix}`.trim(),
+		};
+
+		// If this request requires authorization (allowing non-"authorized" requests for webhooks)
+		if (request.auth === undefined || request.auth) {
+			// If we haven't received a token, throw an error
+			if (!this.#token)
+				return Promise.reject(new Error('Expected token to be set for this request, but none was present'));
+
+			headers.Authorization = `Bot ${this.#token}`;
+		}
+
+		// If a reason was set, set it's appropriate header
+		if (request.reason?.length) {
+			headers['X-Audit-Log-Reason'] = encodeURIComponent(request.reason);
+		}
+
+		let finalBody: RequestInit['body'];
+		let additionalHeaders: Record<string, string> = {};
+
+		if (request.attachments?.length) {
+			const formData = new FormData();
+
+			// Attach all files to the request
+			for (const attachment of request.attachments) {
+				formData.append(attachment.fileName, attachment.rawBuffer, attachment.fileName);
+			}
+
+			// If a JSON body was added as well, attach it to the form data
+			// eslint-disable-next-line no-eq-null
+			if (request.body != null) formData.append('payload_json', JSON.stringify(request.body));
+
+			// Set the final body to the form data
+			finalBody = formData;
+			// Set the additional headers to the form data ones
+			additionalHeaders = formData.getHeaders();
+
+			// eslint-disable-next-line no-eq-null
+		} else if (request.body != null) {
+			// Stringify the JSON data
+			finalBody = JSON.stringify(request.body);
+			// Set the additional headers to specify the content-type
+			additionalHeaders = { 'Content-Type': 'application/json' };
+		}
+
+		const fetchOptions = {
+			agent,
+			body: finalBody,
+			// eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+			headers: { ...(request.headers ?? {}), ...additionalHeaders, ...headers } as Record<string, string>,
+			method: request.method,
+		};
+
+		return Promise.resolve({ url, fetchOptions });
+	}
+
+	/**
+	 * Generates route data for an endpoint:method
+	 * @param endpoint The raw endpoint to generalize
+	 * @param method The HTTP method this endpoint is called without
+	 * @private
+	 */
+	private static generateRouteData(endpoint: string, method: RequestMethod): RouteData {
+		const majorIDMatch = /^\/(?:channels|guilds|webhooks)\/(\d{16,19})/.exec(endpoint);
+
+		// Get the major ID for this route - global otherwise
+		const majorID = majorIDMatch?.[1] ?? 'global';
+
+		const baseRoute = endpoint //
+			// Strip out all IDs
+			.replace(/\d{16,19}/g, ':id')
+			// Strip out reaction as they fall under the same bucket
+			.replace(/\/reactions\/(.*)/, '/reactions/:reaction');
+
+		let exceptions = '';
+
+		// Hard-Code Old Message Deletion Exception (2 week+ old messages are a different bucket)
+		// https://github.com/discordapp/discord-api-docs/issues/1295
+		if (method === 'delete' && baseRoute === '/channels/:id/messages/:id') {
```

Comment:
Would it be beneficial to add a stricter check on the method here?  As a runtime check if a user decides to provide a string for their `REST#request` method instead of using the enum?

Reply:
Unless we want to start validating everything you input in our module, with something like, say, [`ow`](https://npmjs.org/package/ow), if you're manually using `Rest#request` it means you know what you're doing.